### PR TITLE
Partial joints

### PR DESCRIPTION
**Pull request by v-lopez**
Fixes https://github.com/ros-controls/ros2_controllers/issues/31
Fixes https://github.com/ros-controls/ros2_controllers/issues/37

### CODE REVIEW

**Comment by v-lopez** [comment]
@ddengster Could I get your take on these changes?

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 1]
```diff
@@ -630,6 +609,47 @@ void JointTrajectoryController::feedback_setup_callback(
...
+        it.positions.push_back(joint_state->get_position());
+        if (!it.velocities.empty()) {
+          it.velocities.push_back(0.0);
+        }
+        if (!it.accelerations.empty()) {
+          it.accelerations.push_back(0.0);
+        }
+        // TODO(v-lopez) is effort 0.0 valid here? Is it worse than joint_state->get_effort()?
+        if (!it.effort.empty()) {
+          it.effort.push_back(0.0);
```
Comment (ddengster) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:646]: effort implies torque/force, see this [post](https://answers.ros.org/question/331421/ros-control-effort-controllers-explained/?answer=331449#post-id-331449). So setting 0 to torque/force sounds good to me

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 2]
```diff
@@ -630,6 +609,47 @@ void JointTrajectoryController::feedback_setup_callback(
...
+      {
+        // joint found on msg
+        continue;
+      }
+      trajectory_msg->joint_names.push_back(joint_names_[index]);
+
+      const auto & joint_state = registered_joint_state_handles_[index];
+      for (auto & it : trajectory_msg->points) {
+        // Assume hold position with 0 velocity and acceleration for missing joints
+        it.positions.push_back(joint_state->get_position());
```
Comment (ddengster) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:637]: This function is called on a separate thread from the main JTC updating loop, you'll get some kind of race condition if you call `joint_state->get_position()` here. 

`traj_external_point_ptr_->update(msg);` is where we sync up, we could put some kind of coordination or delay it until the first parsing of the message (this is how I did immediate paybacks of trajectory messages where messages are sent with time 0). Perhaps we could use a boolean per joint to indicate whether to retain the last known joint position.

Comment (v-lopez) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:637]: Thanks for noticing this, it is indeed a problem.

Furthermore, On the [topic callback](https://github.com/ros-controls/ros2_controllers/pull/68/files/c910ecc50cb3c53bd6a6c88690f701660de85b43#diff-fbf197065df5790eaa93ff95fc223e8bR332) we update the trajectory without acquiring the `trajectory_mtx_` which is another race condition.

Probably when we receive a trajectory (whether from topic or action) on an "auxiliary thread" we should do some basic checks to ensure that we can accept it, and store it somewhere protected by a mutex.

Then on the next `update` call on the "main thread", we should process it, fill the partial goals and do whatever needs to be done.

Something along these lines I believe would fix these issues.

Comment (v-lopez) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:640]: I've pushed a fix for these issues on this same branch. https://github.com/ros-controls/ros2_controllers/pull/68/commits/188c0e405d8d0db0cdd6c873e41bf9a04d4ffb06

**Pull request merged**